File: Code/State_Machine_Testing/python/generate.py
from functools import cache

from utilities import Utilities
from tokens import Tokens

from variables import Variables
from events import Events

# Used to make passing and failing proteus code
class Generate:
    class Pass:
        def __init__(self):
            self.code = ""
            # States list
            self.num_of_states = Utilities.Random.rand_num(0, len(Tokens().state_names))
            self.states = []
            self.__generate_state_names()

            # Variables
            self.variables = Variables() 

            # Event names mapped with types
            self.events = Events()
    #Pas#---------------------------------------------------------------------------------------------------------------------------------------------#
        # Generate State names
        def __generate_state_names(self):
            # Get statenames list
            state_names = Utilities.Lists.get_list(Tokens().state_names)

            # First item on list. If it's empty, the loop below wont work
            state_name = Utilities.Random.rand_choice(state_names)

            # Puts random states into states list
            for _ in range(self.num_of_states):
                # Makes sure the chosen state name is not already used
                state_name = Utilities.Variable_Handler.avoid_redecloration(state_name, self.states, state_names)
                
                # Adds used statename to list
                self.states.append(state_name)
    #Pas#---------------------------------------------------------------------------------------------------------------------------------------------#
        # Generate an Actor, which makes a statemachine which makes state(s)
        def __generate_actor(self):
            # Get actor names list
            actor_names = Utilities.Lists.get_list(Tokens().actor_names)

            # Picks random name from list
            actor_name = Utilities.Random.rand_choice(actor_names)

            # Generate actor definition
            self.code = f"actor {actor_name} {{\n"

            # Makes statmachine which calls generate_state
            self.code += self.__generate_statemachine()

            # Close actor definitions
            self.code += "}\n"

            # Returns actor code
            return self.code
    #Pas#---------------------------------------------------------------------------------------------------------------------------------------------#
        # Makes statemachine
        def __generate_statemachine(self):

            # Start of statemachine syntax
            code = "\tstatemachine {\n"

            # Picks a state that has been initialized (if states exist)
            # Makes sure the list is not empty
            initial_statename = Utilities.Random.rand_choice(self.states) if self.states else None

            # Randomly decides if we include an initial state 
            if Utilities.Random.rand_choice([True, False]) and initial_statename is not None:
                code += f"\t\tinitial {initial_statename};\n\n"

            # Genrate States
            code += self.__generate_states()

            # Close state machine definitions
            code += "\t}\n"

            # Return State machine syntax
            return code
    #Pas#---------------------------------------------------------------------------------------------------------------------------------------------#
        # Make Random States
        def __generate_states(self):
            # Make and save variables
            self.code = self.variables.generate_variables()

            # Iterating through how many state names we generated
            for i in range(self.num_of_states):
                # state syntax
                self.code += f"\n\t\tstate {self.states[i]} {{\n"

                # Randomly deciding if we want to change variable values in each state
                if Utilities.Random.rand_choice([True, False]):
                    # Randomly select variables to modify
                    num_of_vars = Utilities.Random.rand_num(0, len(self.variables.get_variables()))
                    var_keys = list(self.variables.get_variables().keys())
                    selected_vars = Utilities.Random.rand_samples(var_keys, num_of_vars)
                    
                    # Give new values to the variables
                    for name in selected_vars:
                        value = Utilities.Variable_Handler.make_value(self.variables.get_variables()[name])
                        self.code += f"\t\t\t{name} = {value};\n"
                    self.code += "\t\t}\n"

                # Print blank state
                else:
                    self.code += f"\t\t\t// Add state behavior here;\n"
                    self.code += "\t\t}\n"

            # Returns state code
            return self.code
    #Pas#---------------------------------------------------------------------------------------------------------------------------------------------#
        # Generate random code
        def generate_random_code(self):
            # Events Generated
            self.code += self.events.generate_events()

            # Actor Generated
            self.code += self.__generate_actor()

            # Returns random code
            return self.code
    #Pas#---------------------------------------------------------------------------------------------------------------------------------------------#
#---#---#---------------------------------------------------------------------------------------------------------------------------------------------#
#---#---#---------------------------------------------------------------------------------------------------------------------------------------------#   
    @cache
    class Fail:
        def generate_random_tokens(self, length):
            # Initialize empty string for text 
            random_tokens = ""

            # Line counter
            line_length = 0

            # These rows are for variable names, we want just random tokens
            ignore = Tokens().lists.index(Tokens().punctuation)

            # Random text still has more room
            while len(random_tokens) < length:
                # Picks randomly when to have a space, tab, or \n
                if line_length >= Utilities.Random.rand_num(1, 20):
                    # Add a space, tab, or line break when the line length exceeds the random limit
                    random_tokens += Utilities.Random.rand_choice([" ", "\t", "\n"])

                    # Resets line counter
                    line_length = 0

                # Starts making a new line of random tokens
                else:
                    # Randomly picks tokens, ignores the variable names 
                    selected_list = Utilities.Random.rand_choice(Tokens().lists[:ignore])

                    # Randomly picks item from list, Convert to string
                    selected_item = str(Utilities.Random.rand_choice(selected_list))   

                    # Puts it in the random text
                    random_tokens += selected_item

                    # Adds how many characters were added to line length coutner
                    line_length += len(selected_item)
                    
            # Returns random tokens
            return random_tokens
    #Fai#---------------------------------------------------------------------------------------------------------------------------------------------#

File: Code/State_Machine_Testing/c++/generate.cpp
#include "generate.h"

using std::vector, std::string, std::ofstream;

Generate::Generate()
{
    // Get all code data
    _tokens = _Tokens.getTokens();
    _events = _Events.getEvents();
    _variables = _Variables.getVariables();
    
    // Empty string to write code
    _code = "";
};

string Generate::generateEvents()
{
    // Empty string to write events
    string code = "";

    // Write events
    for (const auto&[name, type] : _events)
    {                     
        code += "event " + name + " {" + type + "};\n";
    }
    code += "\n";

    // Return events
    return code;
}

string Generate::generateVariables()
{
    // Empty string to write variables
    string code = "";

    // Write Variables
    code += "Variables: \n";
    for (const auto&[type, name, value] : _variables)
    {      
        code += type + " " + name + " = " + value + ";\n";
    }
    code += "\n";

    // Write new Variable values
    code += "New Values: \n";
    _Variables.makeNewValues(_variables);
    for (const auto&[type, name, value] : _variables)
    {               
        code += name + " = " + value + ";\n";
    }
    code += "\n";

    // Return variables
    return code;
}

// Can an actor have multiple statemachines?
string Generate::generateActors()
{
    // List of actor names
    vector<string> actorNames = _tokens[Indices::ACTOR_NAMES];

    // Random name from the list
    string randName = actorNames[rand_between(0, actorNames.size() - 1)];

    // Actor syntax
    string actor = "actor " + randName + " {\n";

    // Return syntax with name
    return actor;
}

string Generate::generateStatemachine()
{
    return string();
}

string Generate::generateStates()
{
    return string();
}

// Makes complete proteus code
void Generate::generateRandomCode(const string& file)
{
    // Create and open a text file
    ofstream MyFile(file);

    // Generate events first
    _code += generateEvents();

    // Actor
    _code += generateActors();

    // Variables
    _code += generateVariables();

    // Close actor (for now)
    _code += "}\n";

    //States
    
    // Writes code
    MyFile << _code;

    // Close the file
    MyFile.close();
}

// random randoms to random file randomly
void Generate::generateRandomTokens(const string& file)
{
    // Create and open a text file
    ofstream MyFile(file);

    // Set total limit of written tokens
    int limit = rand_between(10000, 100000);

    // Counter for current line length
    int lineLength = 0;

    // String to collect all random tokens
    string tokens = "";
    
    // Having not hit the limit
    while (tokens.size() < limit)
    {
        // Each code line length is between 1 - 1,000 characters long
        if (lineLength >= rand_between(1, 1000))
        {
            // Randomly Decide if we want a space, tab, or new line
            if (rand_between(0, 1))
            {
                // Randomly choose space, tab, or new line
                switch (rand_between(0, 2))
                {
                    case 0: tokens += " "; break;
                    case 1: tokens += "\t"; break;
                    case 2: tokens += "\n"; break;    
                    default: break;
                }
            }
            // Re-set line legnth counter
            lineLength = 0;
        }

        // all the other tokens that are not space, tab, or new line
        else
        {
            // Select random list before name inputs
            vector<string> selectedList = _tokens[rand_between(0, Indices::PUNCTUATION)];

            // Random string from selected list
            tokens += selectedList[rand_between(0, selectedList.size() - 1)];

            // Length added to compare to limit
            lineLength += tokens.size();
        }
    }
    
    // Write strings to file
    MyFile << tokens;

    // Close the file
    MyFile.close();   
}